mantis-control-plane/mantis-control-plane-server/src/test/java/io/mantisrx/master/resourcecluster/ResourceClusterActorTest.java

```diff
@@ -0,0 +1,165 @@
+        mapper = ResourceClusterTaskExecutorMapper.inMemory();
+    }
+
+    @Before
+    public void setupActor() {
+        final Props props =
+            ResourceClusterActor.props(
+                CLUSTER_ID,
+                heartbeatTimeout,
+                assignmentTimeout,
+                Clock.systemDefaultZone(),
+                rpcService,
+                mantisJobStore);
+
+        resourceClusterActor = actorSystem.actorOf(props);
+        resourceCluster =
+            new ResourceClusterAkkaImpl(
+                resourceClusterActor,
+                Duration.ofSeconds(1),
+                CLUSTER_ID,
+                mapper);
+    }
+
+    @Test
+    public void testRegistration() throws Exception {
+        assertEquals(Ack.getInstance(), resourceCluster.registerTaskExecutor(TASK_EXECUTOR_REGISTRATION).get());
+        assertEquals(ImmutableList.of(TASK_EXECUTOR_ID), resourceCluster.getRegisteredTaskExecutors().get());
+    }
+
+    @Test
+    public void testInitializationAfterRestart() throws Exception {
+        when(mantisJobStore.getTaskExecutor(Matchers.eq(TASK_EXECUTOR_ID))).thenReturn(TASK_EXECUTOR_REGISTRATION);
+        assertEquals(
+            Ack.getInstance(),
+            resourceCluster.initializeTaskExecutor(TASK_EXECUTOR_ID, WORKER_ID).get());
+        assertEquals(ImmutableList.of(TASK_EXECUTOR_ID), resourceCluster.getBusyTaskExecutors().get());
+    }
+
+    @Test
+    public void testGetFreeTaskExecutors() throws Exception {
+        assertEquals(Ack.getInstance(), resourceCluster.registerTaskExecutor(TASK_EXECUTOR_REGISTRATION).get());
+        assertEquals(Ack.getInstance(),
+            resourceCluster
+                .heartBeatFromTaskExecutor(
+                    new TaskExecutorHeartbeat(
+                        TASK_EXECUTOR_ID,
+                        CLUSTER_ID,
+                        TaskExecutorReport.available())).get());
+        assertEquals(TASK_EXECUTOR_ID, resourceCluster.getTaskExecutorFor(MACHINE_DEFINITION, WORKER_ID).get());
+        assertEquals(ImmutableList.of(), resourceCluster.getAvailableTaskExecutors().get());
+        assertEquals(ImmutableList.of(TASK_EXECUTOR_ID), resourceCluster.getRegisteredTaskExecutors().get());
+    }
+
+    @Test
+    public void testAssignmentTimeout() throws Exception {
+        assertEquals(Ack.getInstance(), resourceCluster.registerTaskExecutor(TASK_EXECUTOR_REGISTRATION).get());
+        assertEquals(Ack.getInstance(),
+            resourceCluster
+                .heartBeatFromTaskExecutor(
+                    new TaskExecutorHeartbeat(
+                        TASK_EXECUTOR_ID,
+                        CLUSTER_ID,
+                        TaskExecutorReport.available())).get());
+        assertEquals(TASK_EXECUTOR_ID, resourceCluster.getTaskExecutorFor(MACHINE_DEFINITION, WORKER_ID).get());
+        assertEquals(ImmutableList.of(), resourceCluster.getAvailableTaskExecutors().get());
+        Thread.sleep(2000);
```

Comment:
should this be done using a mock Clock?

Reply:
It would be great if we could. All these `Thread.sleep()` start to add up over time.